File: app/src/main/java/com/wolcano/musicplayer/music/content/slidingpanel/SlidingPanel.java
package com.wolcano.musicplayer.music.content.slidingpanel;

import android.app.Activity;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.squareup.picasso.Picasso;
import com.wolcano.musicplayer.music.R;
import com.wolcano.musicplayer.music.content.Binder;
import com.wolcano.musicplayer.music.widgets.ScaledImageView;
import com.wolcano.musicplayer.music.mvp.listener.OnServiceListener;
import com.wolcano.musicplayer.music.mvp.listener.OnSwipeTouchListener;
import com.wolcano.musicplayer.music.mvp.models.Song;
import com.wolcano.musicplayer.music.provider.RemotePlay;
import com.wolcano.musicplayer.music.widgets.SongCover;
import com.wolcano.musicplayer.music.mvp.listener.Bind;
import com.wolcano.musicplayer.music.ui.activity.MainActivity;
import com.wolcano.musicplayer.music.utils.Utils;
import java.util.concurrent.TimeUnit;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;


public class SlidingPanel implements OnServiceListener,View.OnClickListener {


    private Disposable disposable;
    private Drawable placeholder;
    private Activity activity;


    @Bind(R.id.progressTop1)
    private ProgressBar progressBar;
    @Bind(R.id.line1)
    private TextView line1;
    @Bind(R.id.play)
    private ImageView play;
    @Bind(R.id.line2)
    private TextView line2;
    @Bind(R.id.panel_top1bg)
    private ImageView panelTop1;
    @Bind(R.id.model_imageview)
    private ImageView modelImage;


    public SlidingPanel(View view, Activity activity) {
        Binder.bindIt(this, view);
        this.activity = activity;
        setViews();
    }


    public void setViews(){
        // Marquee TextView
        line1.setHorizontallyScrolling(true);
        line1.setSelected(true);

        play.setColorFilter(Utils.getAccentColor(activity));


        if(RemotePlay.get().getPlayMusic(activity.getApplicationContext())!=null){
            String contentURI = "content://media/external/audio/media/" + RemotePlay.get().getPlayMusic(activity.getApplicationContext()).getSongId() + "/albumart";
            Picasso.get()
                    .load(contentURI)
                    .into(modelImage);
            loadBitmap(RemotePlay.get().getPlayMusic(activity.getApplicationContext()),panelTop1);

        }
        ScaledImageView scaledImageView = new ScaledImageView(activity);
        int width = Resources.getSystem().getDisplayMetrics().widthPixels;
        int height = Resources.getSystem().getDisplayMetrics().heightPixels;

        scaledImageView.imageBitmap = Bitmap.createScaledBitmap(BitmapFactory.decodeResource(activity.getResources(),R.drawable.album_default),width,height,true);
        placeholder = activity.getResources().getDrawable(R.drawable.album_default);
        placeholder.setColorFilter(Utils.getPrimaryColor(activity), PorterDuff.Mode.MULTIPLY);

        progressBar.getProgressDrawable().setColorFilter(Utils.getAccentColor(activity), PorterDuff.Mode.SRC_IN);
        play.setOnClickListener(this);
        panelTop1.setOnTouchListener(new OnSwipeTouchListener(activity) {
            @Override
            public void onSwipeLeft() {
                super.onSwipeLeft();
            }

            @Override
            public void onSwipeRight() {
                super.onSwipeRight();
            }

            @Override
            public void onClick() {
                ((MainActivity) activity).expandPanel();
            }
        });
        onChangeSong(RemotePlay.get().getPlayMusic(activity.getApplicationContext()));
    }
    @Override
    public void onPlayStart() {
        play.setSelected(true);

    }

    @Override
    public void onPlayPause() {
        play.setSelected(false);

    }


    public Disposable getDisposable() {
        return disposable;
    }


    @Override
    public void onProgressChange(int progress) {
        progressBar.setProgress(progress);
    }

    @Override
    public void onBufferingUpdate(int percent) {

    }
    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.play:
                RemotePlay.get().buttonClick(activity);
                break;
        }
    }
    private void setBitmap(Bitmap bitmap,ImageView imageView){
        if(bitmap!=null)
            imageView.setImageBitmap(bitmap);
    }


    @Override
    public void onChangeSong(Song song) {
        if (song == null) {
            line1.setText("");
            line2.setText("");
            progressBar.setProgress(0);
            progressBar.setMax(0);
            Picasso.get()
                    .load(R.drawable.album_art)
                    .into(modelImage);

            loadBitmap(song,panelTop1);
            return;
        }
        String contentURI = "content://media/external/audio/media/" + song.getSongId() + "/albumart";
        Picasso.get()
                .load(contentURI)
                .placeholder(R.drawable.album_art)
                .into(modelImage);

        loadBitmap(song,panelTop1);

        line1.setText(song.getTitle());
        line2.setText(song.getArtist());
        play.setSelected(RemotePlay.get().isPlaying() || RemotePlay.get().isPreparing());
        progressBar.setMax((int) song.getDura());
        progressBar.setProgress((int) RemotePlay.get().getPlayerCurrentPosition());

    }

    private void loadBitmap(Song song, ImageView imageView) {

        Observable<Bitmap> bitmapObservable =
                Observable.fromCallable(() -> SongCover.get().loadBlurred(activity.getApplicationContext(),song)).throttleFirst(500, TimeUnit.MILLISECONDS);
        disposable = bitmapObservable.
                subscribeOn(Schedulers.io()).
                observeOn(AndroidSchedulers.mainThread()).

        subscribeWith(new DisposableObserver<Bitmap>(){


            @Override
            public void onNext(Bitmap bitmap) {
                setBitmap(bitmap,imageView);
            }

            @Override
            public void onError(Throwable e) {
                imageView.setImageDrawable(placeholder);
            }

            @Override
            public void onComplete() {
            }
        });
    }


}

File: app/src/main/java/com/wolcano/musicplayer/music/content/SlidingPanel.kt
package com.wolcano.musicplayer.music.content

import android.app.Activity
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.squareup.picasso.Picasso
import com.wolcano.musicplayer.music.R
import com.wolcano.musicplayer.music.data.model.Song
import com.wolcano.musicplayer.music.listener.Bind
import com.wolcano.musicplayer.music.listener.OnServiceListener
import com.wolcano.musicplayer.music.listener.OnSwipeTouchListener
import com.wolcano.musicplayer.music.provider.RemotePlay.buttonClick
import com.wolcano.musicplayer.music.provider.RemotePlay.getPlayMusic
import com.wolcano.musicplayer.music.provider.RemotePlay.getPlayerCurrentPosition
import com.wolcano.musicplayer.music.provider.RemotePlay.isPlaying
import com.wolcano.musicplayer.music.provider.RemotePlay.isPreparing
import com.wolcano.musicplayer.music.ui.activity.main.MainActivity
import com.wolcano.musicplayer.music.utils.Utils
import com.wolcano.musicplayer.music.widgets.ScaledImageView
import com.wolcano.musicplayer.music.widgets.SongCover
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit

class SlidingPanel(view: View, private val activity: Activity) : OnServiceListener,
    View.OnClickListener {

    var disposable: Disposable? = null
    private var placeholderDrawable: Drawable? = null

    @Bind(R.id.progressTop1)
    private lateinit var progressBar: ProgressBar

    @Bind(R.id.txt_line_1)
    private lateinit var line1: TextView

    @Bind(R.id.img_play)
    private lateinit var play: ImageView

    @Bind(R.id.txt_line_2)
    private lateinit var line2: TextView

    @Bind(R.id.panel_top1bg)
    private lateinit var slidingUpPanelTop1: ImageView

    @Bind(R.id.model_imageview)
    private lateinit var modelImage: ImageView

    init {
        Binder.bindIt(this, view)
        setViews()
    }

    fun setViews() {
        // Marquee TextView
        line1.setHorizontallyScrolling(true)
        line1.isSelected = true
        play.setColorFilter(Utils.getAccentColor(activity))
        if (getPlayMusic(activity.applicationContext) != null) {
            val contentURI = "content://media/external/audio/media/" + getPlayMusic(
                activity.applicationContext
            )?.songId + "/albumart"
            Picasso.get()
                .load(contentURI)
                .into(modelImage)
            loadBitmap(getPlayMusic(activity.applicationContext), slidingUpPanelTop1)
        }
        val scaledImageView = ScaledImageView(activity)
        val width = Resources.getSystem().displayMetrics.widthPixels
        val height = Resources.getSystem().displayMetrics.heightPixels
        scaledImageView.imageBitmap = Bitmap.createScaledBitmap(
            BitmapFactory.decodeResource(
                activity.resources, R.drawable.album_default
            ), width, height, true
        )

        placeholderDrawable = ContextCompat.getDrawable(activity, R.drawable.album_default)
        placeholderDrawable?.setColorFilter(
            Utils.getPrimaryColor(activity),
            PorterDuff.Mode.MULTIPLY
        )
        progressBar.progressDrawable.setColorFilter(
            Utils.getAccentColor(activity),
            PorterDuff.Mode.SRC_IN
        )
        play.setOnClickListener(this)
        slidingUpPanelTop1.setOnTouchListener(object : OnSwipeTouchListener(activity) {
            override fun onClick() {
                super.onClick()
                (activity as MainActivity).expandPanel()
            }
        })

        onChangeSong(getPlayMusic(activity.applicationContext))
    }

    override fun onPlayStart() {
        play.isSelected = true
    }

    override fun onPlayPause() {
        play.isSelected = false
    }

    override fun onProgressChange(progress: Int) {
        progressBar.progress = progress
    }

    override fun onBufferingUpdate(percent: Int) {}

    override fun onClick(v: View) {
        when (v.id) {
            R.id.play -> buttonClick(activity)
        }
    }

    private fun setBitmap(bitmap: Bitmap?, imageView: ImageView) {
        if (bitmap != null) imageView.setImageBitmap(bitmap)
    }

    override fun onChangeSong(song: Song?) {
        if (song == null) {
            line1.text = ""
            line2.text = ""
            progressBar.progress = 0
            progressBar.max = 0
            Picasso.get()
                .load(R.drawable.album_art)
                .into(modelImage)
            loadBitmap(song, slidingUpPanelTop1)
            return
        }
        val contentURI = "content://media/external/audio/media/" + song.songId + "/albumart"
        Picasso.get()
            .load(contentURI)
            .placeholder(R.drawable.album_art)
            .into(modelImage)
        loadBitmap(song, slidingUpPanelTop1)
        line1.text = song.title
        line2.text = song.artist
        play.isSelected = isPlaying() || isPreparing()
        progressBar.max = song.duration.toInt()
        progressBar.progress = getPlayerCurrentPosition().toInt()
    }

    private fun loadBitmap(song: Song?, imageView: ImageView) {
        val bitmapObservable = Observable.fromCallable {
            SongCover.loadBlurred(
                activity.applicationContext, song
            )
        }.throttleFirst(500, TimeUnit.MILLISECONDS)
        disposable =
            bitmapObservable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(object : DisposableObserver<Bitmap?>() {
                    override fun onNext(bitmap: Bitmap) {
                        setBitmap(bitmap, imageView)
                    }

                    override fun onError(e: Throwable) {
                        imageView.setImageDrawable(placeholderDrawable)
                    }

                    override fun onComplete() {}
                })
    }

}